promql/functions.go

```diff
@@ -130,6 +134,193 @@ func extrapolatedRate(vals []parser.Value, args parser.Expressions, enh *EvalNod
+	ms := args[0].(*parser.MatrixSelector)
+	vs := ms.VectorSelector.(*parser.VectorSelector)
+
+	samples := vals[0].(Matrix)[0]
+	rangeStart := enh.Ts - durationMilliseconds(ms.Range+vs.Offset)
+	rangeEnd := enh.Ts - durationMilliseconds(vs.Offset)
+
+	points := samples.Points
+	//** Cannot compute rate with 0 or 1 data points.
+	if len(points) < 2 {
+		return enh.Out
+	}
+	sampledRange := float64(points[len(points)-1].T - points[0].T)
+	averageInterval := sampledRange / float64(len(points)-1) //** msec; if any samples are missing, this will be higher than the scrape interval
+
+	firstPoint := 0
+	// If the point before the range is too far from rangeStart, drop it.
+	if float64(rangeStart-points[0].T) > averageInterval { //** There's 0 slop here, so we could drop the point even if it was scraped 1 msec early
+		//** Repeating the above check for 0 or 1 data points, with +1.
+		if len(points) < 3 { //** This looks like a bug to me. 2 points is enough to measure an increase. -Colin
+			return enh.Out
+		}
+		firstPoint = 1
+		sampledRange = float64(points[len(points)-1].T - points[1].T) //** repeating above code
+		averageInterval = sampledRange / float64(len(points)-2)       //** repeating above code
+	}
+
+	counterCorrection := float64(0.0)
+	lastValue := float64(0.0)
+
+	if isCounter { //** called from rate or increase (not delta)
+		for i := firstPoint; i < len(points); i++ {
+			sample := points[i]
+			if sample.V < lastValue { //** Handle when the counter steps backwards due to process restart
+				counterCorrection += lastValue //** Accumulate the sum of backward steps
+			}
+			lastValue = sample.V
+		}
+	}
+	resultValue := points[len(points)-1].V - points[firstPoint].V + counterCorrection //** last.V - first.V + backward correction
+
+	// Duration between last sample and boundary of range.
+	durationToEnd := float64(rangeEnd - points[len(points)-1].T)
+
+	// If the points cover the whole range (i.e. they start just before the
+	// range start and end just before the range end) adjust the value from
+	// the sampled range to the requested range.
+	if points[firstPoint].T <= rangeStart && durationToEnd < averageInterval {
+		adjustToRange := float64(durationMilliseconds(ms.Range))
+		resultValue *= (adjustToRange / sampledRange)
+	}
+
+	if isRate {
+		resultValue /= ms.Range.Seconds()
+	}
+
+	return append(enh.Out, Sample{
+		Point: Point{V: resultValue},
+	})
+}
+
+// Infers the scrape interval in msec by taking the median interval among the first 25.
+// By using the median, we avoid the highs (late or missed scrape) and the lows
+// (where a late scrape caused the next to be too close).
+func inferScrapeInterval(points []Point) int64 {
+	if len(points) == 0 {
+		panic(errors.Errorf("no points to work on"))
+	}
+
+	// Compute up to 25 time intervals and store them in intervalArray.
+	var intervalArray [25]int64
+	var i int
+	for i = 0; i < (len(points)-1) && i < len(intervalArray); i++ {
+		intervalArray[i] = points[i+1].T - points[i].T
```

Comment:
Yep, the Ruby value system of clear, unabbreviated names is so great. It's infuriating to me that Golang omitted critical features like exceptions and keyword arguments and the ternary operator all in the name of "clarity"...when what all those omissions did was create was boilerplate verbosity... and then, all the Golang code I've read has come up way short in clarity on the variable and attribute names. Which is extra ridiculous since Golang doesn't have arbitrary limits on those names, and modern IDEs will autocomplete those names for you anyway.